### DISCUSSION TITLE
Add state `displaying_app` to media player entity

### DISCUSSION THREAD

**Discussion by emontnemery** [home-assistant/architecture#666] [Entity Models]
### Problem
 Media players such as smart TVs and Google Cast support executing apps which may not support playing media.

### Proposal
Add a new state `displaying_app` to distinguish playing of media from displaying an app.

### Alternative
Document that one of the existing states (`idle`, `on`, `paused`, `playing`) can be used for this purpose.

**Comment by kongo09** [comment]
How is this different from `idle`?

**Reply by balloob** [reply]
No app versus App loaded but nothing playing. In the second case you might want to stop the app to show the home screen.

**Comment by balloob** [comment]
The state `displaying_app` is a bit weird in that it would mean "displaying app and the app is not playing anything". The state would change to "paused"/"playing" the moment the app is loaded.

Why should we introduce a new state for this if we can look at the state attributes to see if `app_id` or `app_name` are set ?

**Reply by TheFes** [reply]
That's what I tried to say here as well: https://github.com/home-assistant/core/pull/66005#issuecomment-1033793266

**Comment by parautenbach** [comment]
How about putting the currently active app in an attribute and leave the state as is?

**Comment by sdrapha** [comment]
The original issue that brought me here is indirectly related. The title of the discussion doesn't fit very well. The issue is that google home devices are capable and should have their volume state available at all times, regardless if the media player is considered off, idle, or displayng_app. So I'm not sure if creating a new state is the answer for that or not. Maybe just the logic of the OFF state shouldn't hide the volume state, that would be enough for this particular issue.

**Reply by erkr** [reply]
In my case I like to read the actual volume in off state as well! This in order to monitor the volume set on my chromecast via google home

**Comment by TheFes** [comment]
It seems that people have issues with their player being shown active when it's `idle`
So an alternative solultion could also be that `idle` devices are not considered active.

So active is:
 - on
 - playing
 - showing_app
 - paused

Inactive:
 - idle
 - off

Where `idle` still allows volume controls.

This might mean that custom_components like `mini_media_card` also need to change their logic.

**Reply by parautenbach** [reply]
What must be state be if an app is shown (being interacted with) and it's playing something (e.g. a music app)?

**Reply by TheFes** [reply]
I would say in that case it should be `playing`

**Comment by TheFes** [comment]
Is there already some consensus on how to approach this?

**Comment by sdrapha** [comment]
My personal vision, if it makes sense for more people:
Trying to organize how it could be

| state | usage  | example | example image |
|:---------:|-----------|---------|---------------|
|  `off` | devices that when turned off, all interactions are disabled | Vizio integration | ![image](https://user-images.githubusercontent.com/25059996/154166570-c02287f0-923a-40eb-8f91-f66b02e5819f.png) Comments: <br /> fine as is, no change needed

| state |   usage  | example | example image | 
|:---------:|-----------|---------|---------------|
| `idle` <br/> OR <br/> `standby` | devices in "off"-ish state that are not doing anything right now, but interactions are enabled, like reading the  current volume and changing it | Google Home devices |  ![image](https://user-images.githubusercontent.com/25059996/154167398-1be58270-9ffd-44f5-80f6-bf7388e39dc3.png) Comments: <br /> Power button, play and mute doesn't do anything, can be ommited. <br/><br/> The icon state color (or card background) should stay as an inactive entity similar to off so it doesn't call to much attention in the UI <br/> <br/> Roku devices are another example in this category, the power button doesn't turn on/off the device in a common way

| state |   usage  | example | example image | 
|:----------:|----------|---------|---------------|
| `connected` <br/> OR <br/> `displaying_app` <br/> OR <br/> `active` | when device is in an active state doing something | Google Home Devices <br/> when active | ![image](https://user-images.githubusercontent.com/25059996/154219203-d483d76a-518e-4df0-bb0e-63ee4e2ecf71.png) <br/> Comments: <br /> Google Home Devices when homeassistant default media player is actively connected, to say something using TTS for example


| state |   usage  | example | example image | 
|:---------:|----------|---------|---------------|
| `playing` |    playing     |    playing media     |![image](https://user-images.githubusercontent.com/25059996/154167283-29865291-a242-4a1d-9860-bb3352af6d22.png) <br/> Comment: <br/> as is, no change needed

**Reply by sdrapha** [reply]
ps. the table format is fine when editting, it always mess it up when I post:
here is a screenshot: 
<details><summary>Click to expand</summary>

![image](https://user-images.githubusercontent.com/25059996/154292995-409d4dbb-a870-4697-9807-4c0143aef848.png)

</details>

**Comment by teamRibcage** [comment]
My preference by far would be for leaving the states of these devices as they are. The current options of `off`, `idle`, `playing`, etc. already allow for an accurate indication of the device's current activity - including if they are displaying an app not "playing" media (for example, connecting an app like Netflix or YouTube to a Chromecast, but NOT starting a video causes a state change from `off` to `idle`. The app name and id of the "displaying app" are shown as attributes of the device). To me, changing this current state scheme doesn't seem to add any functionality, while potentially causes breaking changes to many custom integrations and users' automations based on state.

If there are enough users that see a need to be able to adjust volume while a device is essentially `off`, perhaps just look at changing the way the volume control itself is shown in the front end.

Just my opinion 🙂

**Reply by elupus** [reply]
The added function is to control volume when off. This can't be done currently.

**Reply by parautenbach** [reply]
@elupus this sounds like a special case, perhaps connected to using the universal media player integration. Which hardware are you thinking about that can be off but one can still change the volume (barring old analogue knobs). Would you mind to elaborate?

**Reply by teamRibcage** [reply]
I personally don't see a need to have a volume slider displayed on a device that is `off`, but certainly recognize that others use the software differently than I do. While it's not a persistent slider display, the volume can currently be changed at any time with a `media_player.volume_set` service call. Might further documentation of that functionality help to solve the issue? 🤷‍♀️

**Reply by elupus** [reply]
i don't think volume set is allowed on an off device.

**Reply by teamRibcage** [reply]
You had me second guessing myself @elupus! 🤣
Just double checked, `media_player.volume_set` works on my Google Home and Mini's, even when `off`

**Comment by kongo09** [comment]
Ok, but then the conclusion is:

- no extra state needed
- volume setting is a question of functionality offered by the device
- volume display is a question of feature offered by the card

Thinking more about this, there are actually quite a few devices that are not fully off when off. My receiver and my TV for example go into standby, which still consumes some power and still allows to manipulate a few things of these devices. So if anything, `standby` could be an additional state. But honestly, I think the discussion has shown that this is not needed.

**Reply by sdrapha** [reply]
I'm including your `standby` state into my view of possibilities on my post above

**Comment by sdrapha** [comment]
There is NOT a conclusion yet, that is you trying to fix your point of view. I can see here loud voices against any change at all, that are strong defenders on their own point of view and not open to be in another's person shoes.
Please dont attack me, I'm not defendant of the pull request that was reversed by some of you that got annoyed by the change. Instead I'm just trying to understand ALL point of views and make sense of it. I personnaly don't even care if there is one brand new state or not. But we gotta look at all scenarios and possibilities., as this is an opensource project made to connect ALL things. And here, we are talking architecture of the media_player, not only specifically one integration.

Its so simple and not big work at all to filter out one more state in an automation or custom-component. But the lack of it can't be dealt on the same way, You can't filter IN something that doesn't exist. So that is why I accepted the need for a new state. See my first comment above in the discussion that I didn't came here for it at first, but it makes sense now.

Commenting the above post:

> volume display is a question of feature offered by the card

It is not, because the volume state attribute must exist in the backend in the first place, then the UI can show it or not

> volume setting is a question of functionality offered by the device

Each device and integration has its own capabilities yes. That is why the BASE entity should be prepared for all. Currently, the BASE media_player is limited and missing functionalitty to handle all scenarios.

**Reply by kongo09** [reply]
apologies, the conclusion is my personal conclusion as someone who entered this discussion very open and now settled down on a perspective. Other people might have other conclusions.

I'm not attacking anyone. Not sure why you believe this. If you feel that way, I apologize.

**Reply by sdrapha** [reply]
I didn't say you did. lol. I was just afraid of a bold statement.🤗

**Reply by kongo09** [reply]
thanks. However, to me 

- `off` is truly without power
- `standby` would be essentially `off` but with some limited interaction possibility, like setting volume or reading some states. Or `on` but nothing loaded.
- `idle` would be `on` and potentially having an app loaded, but not really playing anything (which is the core purpose)
- `paused` would be similar to `idle` but with playing being interrupted, so there is a way to continue playing

However, I personally don't think this is necessary, as this is a very technical perspective and not reflecting the user perspective. For a user, `off` and `standby` are often the same and even `idle` and `pause` can be perceived as the same.

**Reply by sdrapha** [reply]
If we have all of them on the backend, all capabilities can be handled correctly, on the technical side.

For the user perspective: that should be handled by the frontend. 
If `off` and `standby` should look the same, that can be done entirely in the frontend, that is their job to make it look right for the user. 

The way a card currently looks shouldn't hold back the backend, and I agree, the UI must be easy to understand and pretty looking, that must be adjusted too, parallel to any of this.

**Comment by TheFes** [comment]
It's not only the setting of the volume, it is the information being available.
I have a script which stores the states of the media_players before sending a TTS for example. 
For those which were playing I have all information available, for those which were not doing anything (so `off`) I don't know the volume, and I'm unable to restore it, unless I turn them on first.

**Reply by sdrapha** [reply]
Exactly! In my post note that I didn't care mentioning service_calls, they work as is. The problem is currently with the states, and states attributes. The lack off it is realy limiting the interactions in many different ways.

**Comment by ScottG489** [comment]
I'm coming to this discussion because I was trying to get volume attributes from a Google Home speaker while it was off and was surprised to find that information wasn't available.

I've tried to follow the discussion and situation best I can. I'm still unclear on a few things. First, why are Google Home devices going into an `off` state? Is this something happening on the Google Home devices themselves? Or does the Google Cast integration just put the speakers in the `off` state after they haven't have anything playing on them for some period? Or is this just something that happens to all media players?

It seems wrong that a device is going into a state that makes it loose information about it automatically.

Also trying to follow what @TheFes is saying, it seems like the reversion (https://github.com/home-assistant/core/pull/66005) should not have been accepted in the first place because @emontnemery's claim was incorrect that it could "no lonegr be distinguished" when a simple query of the attributes would provide that distinguishment.

Also this is more of a general question about this project's method of accepting merges, but it looks like that merge was done with absolutely no discussion? Was there some discussion done out of band somewhere?

**Comment by TheFes** [comment]
@emontnemery has this discussion been given any new insights?
Maybe the idea of @sdrapha could be implemented, where `standby` would be totally inactive, and `idle`  would be something like showing the Photo stream on a Google Nest Hub.

Thanks!
